src/libraries/System.Private.CoreLib/src/System/Threading/PeriodicTimer.cs

```diff
@@ -0,0 +1,223 @@
+            /// </remarks>
+            private PeriodicTimer? _owner;
+            /// <summary>Core of the <see cref="IValueTaskSource{TResult}"/> implementation.</summary>
+            private ManualResetValueTaskSourceCore<bool> _mrvtsc;
+            /// <summary>Cancellation registration for any active <see cref="WaitForNextTickAsync"/> call.</summary>
+            private CancellationTokenRegistration _ctr;
+            /// <summary>Whether the timer has been stopped.</summary>
+            private bool _stopped;
+            /// <summary>Whether there's a pending notification to be received.  This could be due to the timer firing, the timer being stopped, or cancellation being requested.</summary>
+            private bool _signaled;
+            /// <summary>Whether there's a <see cref="WaitForNextTickAsync"/> call in flight.</summary>
+            private bool _activeWait;
+
+            /// <summary>Wait for the next tick of the timer, or for the timer to be stopped.</summary>
+            public ValueTask<bool> WaitForNextTickAsync(PeriodicTimer owner, CancellationToken cancellationToken)
+            {
+                lock (this)
+                {
+                    if (_activeWait)
+                    {
+                        // WaitForNextTickAsync should only be used by one consumer at a time.  Failing to do so is an error.
+                        ThrowHelper.ThrowInvalidOperationException();
+                    }
+
+                    // If the timer has a pending tick or has been stopped, we can complete synchronously.
+                    if (_signaled)
+                    {
+                        // Reset the signal for subsequent consumers, but only if we're not stopped. Since.
+                        // stopping the timer is one way, any subsequent calls should also complete synchronously
+                        // with false, and thus we leave _signaled pinned at true.
+                        if (!_stopped)
+                        {
+                            _signaled = false;
+                        }
+
+                        return new ValueTask<bool>(!_stopped);
+                    }
+
+                    Debug.Assert(!_stopped, "Unexpectedly stopped without _signaled being true.");
+
+                    // If cancellation has already been requested, short-circuit.
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return ValueTask.FromCanceled<bool>(cancellationToken);
+                    }
+
+                    // Set up for the wait and return a task that will be signaled when the
+                    // timer fires, stop is called, or cancellation is requested.
+                    _owner = owner;
+                    _activeWait = true;
+                    _ctr = cancellationToken.UnsafeRegister(static (state, cancellationToken) => ((State)state!).Signal(cancellationToken: cancellationToken), this);
```

Comment:
Check `cancellationToken.CanBeCancelled` here

Reply:
UnsafeRegister itself does that check.